cpp/tests/pagerank/pagerank_mg_test.cpp

```diff
@@ -0,0 +1,257 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <utilities/base_fixture.hpp>
+#include <utilities/mg_test_utilities.hpp>
+#include <utilities/test_utilities.hpp>
+
+#include <algorithms.hpp>
+#include <partition_manager.hpp>
+
+#include <raft/comms/mpi_comms.hpp>
+
+#include <gtest/gtest.h>
+
+#include <random>
+
+////////////////////////////////////////////////////////////////////////////////
+// Test param object. This defines the input and expected output for a test, and
+// will be instantiated as the parameter to the tests defined below using
+// INSTANTIATE_TEST_CASE_P()
+//
+typedef struct Pagerank_Testparams_t {
+  std::string graph_file_full_path{};
+  double personalization_ratio{0.0};
+  bool test_weighted{false};
+
+  Pagerank_Testparams_t(std::string const& graph_file_path,
+                        double personalization_ratio,
+                        bool test_weighted)
+    : personalization_ratio(personalization_ratio), test_weighted(test_weighted)
+  {
+    if ((graph_file_path.length() > 0) && (graph_file_path[0] != '/')) {
+      graph_file_full_path = cugraph::test::get_rapids_dataset_root_dir() + "/" + graph_file_path;
+    } else {
+      graph_file_full_path = graph_file_path;
+    }
+  };
+} Pagerank_Testparams_t;
+
+////////////////////////////////////////////////////////////////////////////////
+// Parameterized test fixture, to be used with TEST_P().  This defines common
+// setup and teardown steps as well as common utilities used by each E2E MG
+// test.  In this case, each test is identical except for the inputs and
+// expected outputs, so the entire test is defined in the run_test() method.
+//
+class Pagerank_E2E_MG_Testfixture_t : public ::testing::TestWithParam<Pagerank_Testparams_t> {
+ public:
+  Pagerank_E2E_MG_Testfixture_t() {}
+
+  // Run once for the entire fixture
+  //
+  // FIXME: consider a ::testing::Environment gtest instance instead, as done
+  // by cuML in test_opg_utils.h
+  static void SetUpTestCase()
```

Comment:
I think this is in your TODO list for the PR but... is there a good way to make this more reusable?  We'll want to have dozens of tests that will have the same setup/teardown steps for the MG cluster.  It would be nice to either inherit from a common parent or instantiate a common data member that we call methods from.

I think the goal should be to be able to create a new unit test for a different algorithm only having to worry about what's unique about that algorithm... not having to remember the proper sequence of things to initialize/setup.

Reply:
I agree, I'll make an attempt at a more reusable MG test base class in the next commit.

Reply:
I added a pretty trivial base class as a starting point. I'll mark this as resolved but if you think it should be updated further, feel free to unresolve and I'll take a look.